### DISCUSSION TITLE
Could the various `Psl\Type` functions be considered "pure"?

### DISCUSSION THREAD

**Discussion by gsteel** [azjezz/psl#443] [Q&A] [answered]
It looks like `Psl\Type\literal_scalar(42)` could be considered `@psalm-pure`, and the same could be said of all the sibling functions (?)

If so, would it be worth sending in a patch for updated doc-blocks?

The cascade of Psalm issues would be huge… I guess all of the `Psl\Type\Internal` classes would need to be marked with `@psalm-immutable`, and then any other used functions with more `@psalm-pure`.

Is it worth it?

**Comment by veewee** [comment, marked as answer]
It is not possible for all types to be marked as pure.
Once iterables are iterated, there might be side-effects (e.g. inside generators)

See:

* https://github.com/azjezz/psl/issues/99
* https://github.com/azjezz/psl/pull/101#pullrequestreview-525812873

**Reply by gsteel** [reply]
That makes sense. Thanks for the detail :)
